doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,348 @@
+of their GUI-only authors. Every additional `*w` version of an application is an additional maintenance burden and
+source of cognitive overhead\[2\] for users.
+
+On the other side, you have mostly-GUI applications that want to print output to a console **if there is one
+connected**.
+
+These applications are still primarily GUI-driven, but they might support arguments like `/?` or `--help`. They only
+need a console when they need to print out some text. Sometimes they'll allocate their own console (which opens a new
+window) to display in, and sometimes they'll reattach to the originating console. VSCode does the latter, and so when
+you run `code` from CMD, and then `exit` CMD, your console window sticks around because VSCode is still attached to it.
+It will never print anything, and your only option is to close it.
+
+There's another risk risk in reattaching, too. Given that the shell decides whether to wait based on the subsystem
+field, GUI subsystem applications that reattach to their owning consoles *just to print some text* end up stomping on
+the output of any shell that doesn't wait for them:
+
+```
+C:\> application --help
+
+application - the interesting application
+C:\> Usage: application [OPTIONS] ...
+```
+
+> _(the prompt is interleaved with the output)_
+
+## Solution Design
+
+I propose that we introduce a fusion manifest field, **consoleAllocationPolicy**, with the following values:
+
+* _absent_
+* `detached`
+
+This field allows an application to disable the automatic allocation of a console, regardless of the [process creation flags]
+passed to [`CreateProcess`] and its subsystem value.
+
+It would look (roughly) like this:
+
+```xml
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+  <application>
+    <windowsSettings>
+      <consoleAllocationPolicy xmlns="http://schemas.microsoft.com/SMI/20XX/WindowsSettings">detached</consoleAllocationPolicy>
+    </windowsSettings>
+  </application>
+</assembly>
+```
+
+The effects of this field will only apply to binaries in the `IMAGE_SUBSYSTEM_WINDOWS_CUI` subsystem, as it pertains to
+the particulars of their console allocation.
+
+**All console inheritance will proceed as normal.** Since this field takes effect only in the absence of console
+inheritance, CUI applications will still be able to run inside an existing console session.
+
+| policy     | behavior                                                                                                            |
+| -          | -                                                                                                                   |
+| _absent_   | _default behavior_                                                                                                  |
+| `detached` | The new process is not attached to a console session (similar to `DETACHED_PROCESS`) unless one was inherited.      |

```

Comment:
A potentially better name might be "manual" since "detached" sounds a bit like you aren't connected to the parent console anymore.

Reply:
What seems to be implemented by the spec is an inherit-only policy. That's similar to a POSIX terminal app, which depends on inheriting standard I/O files associated with a tty. It's just that on Windows instead of inheriting a standard file triple, a process inherits a file quadruple, which includes a console connection as well as stdin, stdout, and stderr. Anyway, maybe "posix" would be a succinct and clear name for the policy, or "inherit_only". In contrast, the default "windows" console allocation policy inherits or allocates by default, except when the spawning process overrides the behavior by specifying one of the process creation flags `CREATE_NEW_CONSOLE`, `CREATE_NO_WINDOW`, or `DETACHED_PROCESS`.